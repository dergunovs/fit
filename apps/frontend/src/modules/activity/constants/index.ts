import { TActivityChartType } from 'fitness-tracker-contracts';

export const URL_ACTIVITY = '/activities';
export const URL_ACTIVITY_CREATE = `${URL_ACTIVITY}/create`;
export const URL_ACTIVITY_EDIT = `${URL_ACTIVITY}/edit`;

export const URL_ACTIVITY_ADMIN = '/activities_admin';
export const URL_ACTIVITY_ADMIN_EDIT = `${URL_ACTIVITY_ADMIN}/edit`;

export const ACTIVITY_LIST_HEADERS = [{ title: 'Создано' }, { title: 'Пользователь' }];

export const CHART_TYPES: { title: string; value: TActivityChartType }[] = [
  { title: 'Занятия', value: 'activity' },
  { title: 'Подходы', value: 'set' },
  { title: 'Повторы', value: 'repeat' },
  { title: 'Мышцы', value: 'muscle' },
];

export const ACTIVITY_STATISTICS_GAP = 30;
