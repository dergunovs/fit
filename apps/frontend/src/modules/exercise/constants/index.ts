export const URL_EXERCISE = '/exercises';
export const URL_EXERCISE_CREATE = `${URL_EXERCISE}/create`;
export const URL_EXERCISE_EDIT = `${URL_EXERCISE}/edit`;

export const EXERCISE_REPEATS_OPTIONS = [8, 10, 12, 15, 20];
export const EXERCISE_REPEATS_DEFAULT = 12;

export const EXERCISE_STATISTICS_HEADERS = [{ title: 'Упражнение' }, { title: 'Подходы' }, { title: 'Повторы' }];
export const MUSCLE_GROUP_STATISTICS_HEADERS = [{ title: 'Мышцы' }, { title: 'Подходы' }, { title: 'Повторы' }];
export const EXERCISE_LIST_HEADERS = [{ title: 'Название' }];

export const EXERCISE_DELETED_TITLE = 'Упражнение удалено';
