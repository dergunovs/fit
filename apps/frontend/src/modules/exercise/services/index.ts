import { ComputedRef, Ref } from 'vue';
import {
  API_EXERCISE,
  API_EXERCISE_ALL,
  API_EXERCISE_CUSTOM,
  TGetExercisesDTO,
  TGetExercisesCustomDTO,
  TGetExercisesQueryDTO,
  TGetExerciseDTO,
  TPostExerciseDTO,
  TPostExerciseDataDTO,
  TUpdateExerciseDTO,
  TUpdateExerciseDataDTO,
  TDeleteExerciseDTO,
} from 'fitness-tracker-contracts';
import { useMutation, useQuery, api } from 'mhz-helpers';

export const exerciseService = {
  getMany: (page: Ref<number>) => {
    return useQuery({
      queryKey: [API_EXERCISE, page],
      queryFn: async () => {
        const params: TGetExercisesQueryDTO = { page: page.value };

        const { data } = await api.get<TGetExercisesDTO>(API_EXERCISE, { params });

        return data;
      },
    });
  },

  getAll: () => {
    return useQuery({
      queryKey: [API_EXERCISE, API_EXERCISE_ALL],
      queryFn: async () => {
        const { data } = await api.get<TGetExercisesDTO>(API_EXERCISE_ALL);

        return data.data;
      },
    });
  },

  getCustom: () => {
    return useQuery({
      queryKey: [API_EXERCISE, API_EXERCISE_CUSTOM],
      queryFn: async () => {
        const { data } = await api.get<TGetExercisesCustomDTO>(API_EXERCISE_CUSTOM);

        return data.data;
      },
    });
  },

  getOne: (options: object, id?: ComputedRef<string | undefined>) => {
    return useQuery({
      queryKey: [API_EXERCISE, id],
      queryFn: async () => {
        if (!id?.value) return null;

        const { data } = await api.get<TGetExerciseDTO>(`${API_EXERCISE}/${id.value}`);

        return data.data;
      },
      ...options,
    });
  },

  create: (options: object) => {
    return useMutation({
      mutationKey: [API_EXERCISE],
      mutationFn: async (formData: TPostExerciseDataDTO) => {
        const { data } = await api.post<TPostExerciseDTO>(API_EXERCISE, formData);

        return data;
      },
      ...options,
    });
  },

  update: (options: object) => {
    return useMutation({
      mutationKey: [API_EXERCISE],
      mutationFn: async (formData: TUpdateExerciseDataDTO) => {
        const { data } = await api.patch<TUpdateExerciseDTO>(`${API_EXERCISE}/${formData._id}`, formData);

        return data;
      },
      ...options,
    });
  },

  delete: (options: object) => {
    return useMutation({
      mutationKey: [API_EXERCISE],
      mutationFn: async (id: string) => {
        const { data } = await api.delete<TDeleteExerciseDTO>(`${API_EXERCISE}/${id}`);

        return data;
      },
      ...options,
    });
  },
};
