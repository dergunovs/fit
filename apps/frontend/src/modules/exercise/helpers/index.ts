import {
  IExercise,
  IExerciseChoosen,
  IExerciseDone,
  IExerciseStatistics,
  IUser,
  IUserEquipment,
} from 'fitness-tracker-contracts';
import { createTempId } from 'mhz-helpers';

import { ITimelineStep } from '@/activity/interface';
import { getWeightsForUserEquipment } from '@/equipment/helpers';

export function getUserEquipmentParams(exercise: IExercise, user?: IUser) {
  const isExerciseHasEquipment = !!exercise.equipment;
  const isExerciseHasEquipmentForWeight = !!exercise.equipmentForWeight?.length;
  const isWeightsRequired = !!exercise.isWeightsRequired;

  const isUserHasEquipment = !!user?.equipments?.some(
    (equipment) => equipment.equipment?._id?.toString() === exercise.equipment?._id?.toString()
  );

  const isUserHasEquipmentForWeight = !!user?.equipments?.some((equipment) =>
    exercise.equipmentForWeight?.some(
      (equipmentForWeight) => equipmentForWeight._id?.toString() === equipment.equipment?._id?.toString()
    )
  );

  return {
    isExerciseHasEquipment,
    isExerciseHasEquipmentForWeight,
    isWeightsRequired,
    isUserHasEquipment,
    isUserHasEquipmentForWeight,
  };
}

export function isUserEquipmentMatches(exercise: IExercise, user?: IUser) {
  let result = false;

  const {
    isExerciseHasEquipment,
    isExerciseHasEquipmentForWeight,
    isWeightsRequired,
    isUserHasEquipment,
    isUserHasEquipmentForWeight,
  } = getUserEquipmentParams(exercise, user);

  if (!isExerciseHasEquipment && !isExerciseHasEquipmentForWeight) {
    result = true;
  } else if (isUserHasEquipment && !isWeightsRequired) {
    result = true;
  } else if (!isExerciseHasEquipment && isExerciseHasEquipmentForWeight && isUserHasEquipmentForWeight) {
    result = true;
  } else if (!isExerciseHasEquipment && !isWeightsRequired) {
    result = true;
  } else if (
    isExerciseHasEquipment &&
    isExerciseHasEquipmentForWeight &&
    isUserHasEquipment &&
    isUserHasEquipmentForWeight
  ) {
    result = true;
  }

  return result;
}

export function getAverageDuration(exercise: IExerciseStatistics, type: 'set' | 'repeat') {
  return type === 'set'
    ? `${((exercise.averageDuration * exercise.repeats) / exercise.sets || 0).toFixed(1)}`
    : `${exercise.averageDuration.toFixed(1)}`;
}

export function getAverageRepeatsInSet(repeats?: number, sets?: number) {
  return repeats && sets ? Math.round(repeats / sets) : 0;
}

export function isPrevExerciseSame(exercises: IExerciseDone[], index: number, id?: string) {
  return id && exercises[index - 1] ? id === exercises[index - 1].exercise?._id : false;
}

export function isSetCreatable(choosenExercises: IExerciseChoosen[], index: number, id?: string) {
  return !isPrevExerciseSame(choosenExercises, index, id) && index > 0 && index + 1 === choosenExercises.length;
}

export function generateTimeline(exercises: IExerciseDone[], start: Date | string | null, ratio: number) {
  const allSteps: ITimelineStep[] = [];

  exercises.forEach((exercise, index) => {
    if (!exercise.dateUpdated || !exercise.duration || !start) return;

    allSteps.push({
      left:
        index === 0
          ? 0
          : (new Date(exercise.dateUpdated).getTime() - exercise.duration * 1000 - new Date(start).getTime()) / ratio,
      right:
        index === exercises.length - 1 && exercises.length !== 1
          ? 0
          : (new Date(exercise.dateUpdated).getTime() - new Date(start).getTime()) / ratio,
      type: 'exercise',
    });
  });

  allSteps.forEach((step, index) => {
    if (!allSteps[index - 1]?.right) return;

    allSteps.push({ left: allSteps[index - 1].right, right: step.left, type: 'rest' });
  });

  return allSteps.sort((a, b) => a.left - b.left);
}

export function filterExercisesByTitleAndMuscle(exercises: IExercise[], title: string, muscle?: string, user?: IUser) {
  return exercises.filter((exercise) => {
    const isEquipmentMatches = isUserEquipmentMatches(exercise, user);

    const titleFilter =
      (exercise.title.toLowerCase().includes(title.toLocaleLowerCase()) ||
        exercise.title_en?.toLowerCase().includes(title.toLocaleLowerCase())) &&
      isEquipmentMatches;
    const muscleFilter = exercise.muscles?.some((group) => group._id === muscle) && isEquipmentMatches;

    return muscle ? muscleFilter && titleFilter : titleFilter;
  });
}

export function getAvailableExerciseWeights(exercise: IExercise, user: IUser) {
  const equipments = user.equipments?.filter((equipment) =>
    exercise.equipmentForWeight?.some((eq) => eq._id === equipment.equipment?._id)
  );

  if (!equipments?.length) return [];

  const weights = getWeightsForUserEquipment(equipments);

  return exercise.isWeightsRequired ? weights : [0, ...weights];
}

export function getExercisesToChooseDefaultWeight(exercises: IExercise[], userEquipments?: IUserEquipment[]) {
  const equipmentsForWeight = userEquipments?.filter((eq) => eq.weights?.length);
  const exercisesWithWeight = exercises.filter((ex) => ex.isWeights);

  const tableData = exercisesWithWeight.map((ex) => {
    const exerciseEquipment = ex.equipmentForWeight
      ?.filter((eq) => equipmentsForWeight?.some((equipment) => equipment.equipment?._id === eq._id))
      .map((exToMap) => exToMap._id);

    const availableEquipment = equipmentsForWeight?.filter((eq) =>
      exerciseEquipment?.some((equipmentId) => equipmentId === eq.equipment?._id)
    );

    const options = getWeightsForUserEquipment(availableEquipment);

    return {
      _id: ex._id,
      title: ex.title,
      title_en: ex.title_en,
      options: ex.isWeightsRequired ? options : [0, ...options],
    };
  });

  return tableData.filter((ex) => ex.options.length > 1);
}

export function getDefaultExerciseWeight(exercise: IExercise, user: IUser, weights?: number[]) {
  let weight = weights?.length && exercise.isWeightsRequired ? weights[0] : 0;

  if (user.defaultWeights && exercise._id && Object.hasOwn(user.defaultWeights, exercise._id)) {
    weight = user.defaultWeights[exercise._id];
  }

  return weight;
}

export function addSetToExercises(exercises: IExerciseChoosen[]) {
  const set = exercises.slice(-2).map((exercise) => {
    return { ...exercise, _id: createTempId() };
  });

  return [...exercises, ...set];
}

export function updateExercisesIndex(exercises: IExerciseChoosen[], index: number) {
  if (index <= 0 || index >= exercises.length) return exercises;

  const updatedExercises = [...exercises];
  const exerciseToMove = updatedExercises[index];

  updatedExercises.splice(index, 1);
  updatedExercises.splice(index - 1, 0, exerciseToMove);

  return updatedExercises;
}

export function updateExerciseField<T extends keyof IExerciseDone>(
  exercises: IExerciseChoosen[],
  field: T,
  value: IExerciseDone[T],
  id?: string
): IExerciseChoosen[] {
  return exercises.map((exercise) => (exercise._id === id ? { ...exercise, [field]: value } : exercise));
}

export function generateExerciseChoosen(repeats: number, weight?: number, exercise?: IExercise): IExerciseChoosen {
  return {
    _id: createTempId(),
    exercise: {
      _id: exercise?._id || '',
      title: exercise?.title || '',
      title_en: exercise?.title_en || '',
      muscles: exercise?.muscles || [],
      isWeights: exercise?.isWeights || false,
      isWeightsRequired: exercise?.isWeightsRequired || false,
      equipmentForWeight: exercise?.equipmentForWeight,
    },
    repeats,
    weight,
  };
}
