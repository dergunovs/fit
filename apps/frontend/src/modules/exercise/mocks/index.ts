import { vi } from 'vitest';
import {
  TDeleteExerciseDTO,
  TGetExercisesDTO,
  TPostExerciseDataDTO,
  TPostExerciseDTO,
  TUpdateExerciseDataDTO,
  TUpdateExerciseDTO,
} from 'fitness-tracker-contracts';

import { exerciseService } from '@/exercise/services';
import { mockMutationReply, mockQueryReply } from '@/common/mocks';
import { EXERCISE_FIXTURE, EXERCISE_FIXTURE_CUSTOM, EXERCISES_FIXTURE } from '@/exercise/fixtures';
import { IOnSuccess } from '@/common/interface';

const spyGetExercise = vi.spyOn(exerciseService, 'getOne').mockReturnValue(mockQueryReply(EXERCISE_FIXTURE));

const mockGetExercisesData: TGetExercisesDTO = { data: EXERCISES_FIXTURE, total: EXERCISES_FIXTURE.length };

const spyGetExercises = vi
  .spyOn(exerciseService, 'getMany')
  .mockImplementation(() => mockQueryReply(mockGetExercisesData));

const spyGetExercisesAll = vi
  .spyOn(exerciseService, 'getAll')
  .mockImplementation(() => mockQueryReply(EXERCISES_FIXTURE));

const spyGetExercisesCustom = vi
  .spyOn(exerciseService, 'getCustom')
  .mockImplementation(() => mockQueryReply([EXERCISE_FIXTURE_CUSTOM]));

const spyCreateExercise = vi.fn();
const spyUpdateExercise = vi.fn();
const spyDeleteExercise = vi.fn();

const mockOnSuccess: IOnSuccess = {
  create: undefined,
  update: undefined,
  delete: undefined,
};

vi.spyOn(exerciseService, 'create').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) mockOnSuccess.create = options.onSuccess;

  return mockMutationReply<TPostExerciseDTO, TPostExerciseDataDTO>(spyCreateExercise);
});

vi.spyOn(exerciseService, 'update').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) mockOnSuccess.update = options.onSuccess;

  return mockMutationReply<TUpdateExerciseDTO, TUpdateExerciseDataDTO>(spyUpdateExercise);
});

vi.spyOn(exerciseService, 'delete').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) mockOnSuccess.delete = options.onSuccess;

  return mockMutationReply<TDeleteExerciseDTO, string>(spyDeleteExercise);
});

export {
  spyGetExercise,
  spyGetExercises,
  spyGetExercisesCustom,
  spyGetExercisesAll,
  spyCreateExercise,
  spyUpdateExercise,
  spyDeleteExercise,
  mockOnSuccess,
  mockGetExercisesData,
};
