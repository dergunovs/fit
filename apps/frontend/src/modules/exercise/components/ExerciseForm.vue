<template>
  <div>
    <UiFlex @submit.prevent="submit" tag="form" column gap="24" align="flex-start" data-test="exercise-form">
      <UiField label="Название" isRequired :error="error('title')">
        <UiInput v-model="formData.title" data-test="exercise-form-title" />
      </UiField>

      <UiField label="Описание">
        <UiEditor v-model="formData.description" data-test="exercise-form-description" />
      </UiField>

      <UiCheckbox v-model="formData.isWeights" label="Наличие веса" data-test="exercise-form-is-weights" />

      <UiCheckbox
        v-if="formData.isWeights"
        v-model="formData.isWeightsRequired"
        label="Обязательность веса"
        data-test="exercise-form-is-weights-required"
      />

      <UiField label="Используемое оборудование">
        <UiSelect v-model="formData.equipment" :options="filterEquipmentByWeights(equipments)" isClearable />
      </UiField>

      <UiFlex v-if="formData.isWeights" column>
        <div>Возможное оборудование для веса</div>

        <UiCheckbox
          v-for="equipment in filterEquipmentByWeights(equipments, true)"
          :key="equipment.title"
          :modelValue="choosenEquipmentForWeight.some((eq) => equipment.title === eq.title)"
          @update:modelValue="updateEquipment(equipment, !!$event)"
          :label="equipment.title"
          data-test="exercise-form-equipment-for-weight"
        />
      </UiFlex>

      <UiFlex column>
        <div>Задействованные мышцы</div>

        <UiCheckbox
          v-for="muscle in muscles"
          :key="muscle._id"
          :modelValue="choosenMuscles.some((group) => group._id === muscle._id)"
          @update:modelValue="updateMuscles(muscle, !!$event)"
          :label="muscle.title"
          data-test="exercise-form-muscle-groups"
        />
      </UiFlex>

      <FormButtons
        :id="props.exercise?._id"
        :isLoading="isLoadingPost || isLoadingUpdate"
        :isEdit="props.isEdit"
        :isEmitCancel="props.isDisableRedirect"
        @delete="(id) => mutateDelete(id)"
        @cancel="emit('hide')"
        data-test="exercise-form-buttons"
      />
    </UiFlex>
  </div>
</template>

<script setup lang="ts">
import { ref, onMounted } from 'vue';
import { useRouter } from 'vue-router';
import { UiField, UiInput, UiCheckbox, toast, UiSelect, UiFlex, UiEditor } from 'mhz-ui';
import { useQueryClient, useValidator, required, clone } from 'mhz-helpers';
import { IExercise, IMuscle, IEquipment, API_EXERCISE, API_ACTIVITY_STATISTICS } from 'fitness-tracker-contracts';

import FormButtons from '@/common/components/FormButtons.vue';

import { exerciseService } from '@/exercise/services';
import { equipmentService } from '@/equipment/services';
import { muscleService } from '@/muscle/services';
import { URL_EXERCISE } from '@/exercise/constants';
import { filterEquipmentByWeights } from '@/equipment/helpers';

interface IProps {
  exercise?: IExercise | null;
  isEdit?: boolean;
  isDisableRedirect?: boolean;
}

const props = defineProps<IProps>();
const emit = defineEmits<{ hide: [] }>();

const queryClient = useQueryClient();

const router = useRouter();

const formData = ref<IExercise>({
  title: '',
  description: '',
  muscles: [],
  isWeights: false,
  isWeightsRequired: false,
  equipment: undefined,
  equipmentForWeight: [],
});

const choosenMuscles = ref<IMuscle[]>([]);
const choosenEquipmentForWeight = ref<IEquipment[]>([]);

const { data: equipments } = equipmentService.getAll();
const { data: muscles } = muscleService.getAll();

function updateMuscles(muscle: IMuscle, isChecked: boolean) {
  choosenMuscles.value = isChecked
    ? [...choosenMuscles.value, muscle]
    : choosenMuscles.value.filter((current) => current._id !== muscle._id);

  formData.value.muscles = [...choosenMuscles.value];
}

function updateEquipment(equipment: IEquipment, isChecked: boolean) {
  choosenEquipmentForWeight.value = isChecked
    ? [...choosenEquipmentForWeight.value, equipment]
    : choosenEquipmentForWeight.value.filter((current) => current.title !== equipment.title);

  formData.value.equipmentForWeight = [...choosenEquipmentForWeight.value];
}

const { mutate: mutatePost, isPending: isLoadingPost } = exerciseService.create({
  onSuccess: async () => {
    await queryClient.refetchQueries({ queryKey: [API_EXERCISE] });
    await queryClient.refetchQueries({ queryKey: [API_ACTIVITY_STATISTICS] });
    toast.success('Упражнение добавлено');
    if (!props.isDisableRedirect) router.push(URL_EXERCISE);
    emit('hide');
  },
});

const { mutate: mutateUpdate, isPending: isLoadingUpdate } = exerciseService.update({
  onSuccess: async () => {
    await queryClient.refetchQueries({ queryKey: [API_EXERCISE] });
    await queryClient.refetchQueries({ queryKey: [API_ACTIVITY_STATISTICS] });
    toast.success('Упражнение обновлено');
    emit('hide');
  },
});

const { mutate: mutateDelete } = exerciseService.delete({
  onSuccess: async () => {
    await queryClient.refetchQueries({ queryKey: [API_EXERCISE] });
    await queryClient.refetchQueries({ queryKey: [API_ACTIVITY_STATISTICS] });
    toast.success('Упражнение удалено');
    if (!props.isDisableRedirect) router.push(URL_EXERCISE);
    emit('hide');
  },
});

const { error, isValid } = useValidator(formData, {
  title: [required()],
});

function submit() {
  if (!isValid()) return;

  if (props.exercise?._id) {
    mutateUpdate(formData.value);
  } else {
    mutatePost(formData.value);
  }
}

onMounted(() => {
  if (props.exercise) {
    formData.value = clone(props.exercise);

    if (formData.value.muscles?.length) choosenMuscles.value = [...formData.value.muscles];
    if (formData.value.equipmentForWeight?.length)
      choosenEquipmentForWeight.value = [...formData.value.equipmentForWeight];
  }
});
</script>
