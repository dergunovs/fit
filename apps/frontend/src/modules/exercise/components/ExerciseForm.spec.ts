import { nextTick } from 'vue';
import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { VueWrapper, enableAutoUnmount } from '@vue/test-utils';
import { API_ACTIVITY_STATISTICS, API_EXERCISE } from 'fitness-tracker-contracts';
import { UiCheckbox } from 'mhz-ui';
import { dataTest } from 'mhz-helpers';

import ExerciseForm from './ExerciseForm.vue';
import FormButtons from '@/common/components/FormButtons.vue';

import { wrapperFactory } from '@/common/test';
import { mockOnSuccess, spyCreateExercise, spyUpdateExercise, spyDeleteExercise } from '@/exercise/mocks';
import { spyRefetchQueries, spyRouterPush, spyToastSuccess, mockIsValid, spyRemoveQueries } from '@/common/mocks';
import { URL_EXERCISE } from '@/exercise/constants';
import { EXERCISE_FIXTURE } from '@/exercise/fixtures';
import { spyGetEquipments } from '@/equipment/mocks';
import { filterEquipmentByWeights } from '@/equipment/helpers';
import { EQUIPMENTS_FIXTURE } from '@/equipment/fixtures';
import { MUSCLES_FIXTURE } from '@/muscle/fixtures';
import { spyGetMuscles } from '@/muscle/mocks';

const TITLE = 'Упражнение';
const TITLE_EN = 'Exercise';
const DESCRIPTION = 'Описание упражнения';
const DESCRIPTION_EN = 'Exercise descriptions';

const form = dataTest('exercise-form');
const formTitle = dataTest('exercise-form-title');
const formTitleEn = dataTest('exercise-form-title-en');
const formDescription = dataTest('exercise-form-description');
const formDescriptionEn = dataTest('exercise-form-description-en');
const formIsWeights = dataTest('exercise-form-is-weights');
const formIsWeightsRequired = dataTest('exercise-form-is-weights-required');
const formEquipmentForWeight = dataTest('exercise-form-equipment-for-weight');
const formMuscleGroups = dataTest('exercise-form-muscle-groups');
const formButtons = dataTest('exercise-form-buttons');

const wrapperWithExercise: VueWrapper<InstanceType<typeof ExerciseForm>> = wrapperFactory(ExerciseForm, {
  exercise: EXERCISE_FIXTURE,
  isEdit: true,
});

let wrapper: VueWrapper<InstanceType<typeof ExerciseForm>>;

beforeEach(() => {
  wrapper = wrapperFactory(ExerciseForm);
});

enableAutoUnmount(afterEach);

describe('ExerciseForm', async () => {
  it('exists', async () => {
    expect(wrapper.findComponent(ExerciseForm)).toBeTruthy();
  });

  it('matches snapshot', async () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('gets equipment', async () => {
    expect(spyGetEquipments).toBeCalledTimes(1);
  });

  it('gets muscles', async () => {
    expect(spyGetMuscles).toBeCalledTimes(1);
  });

  it('shows equipment for weights', async () => {
    await wrapper.findComponent(formIsWeights).setValue(true);

    expect(wrapper.findAll(formEquipmentForWeight).length).toBe(
      filterEquipmentByWeights(EQUIPMENTS_FIXTURE, true).length
    );

    expect(wrapper.find(formEquipmentForWeight).attributes('label')).toBe(
      filterEquipmentByWeights(EQUIPMENTS_FIXTURE, true)[0].title
    );
  });

  it('uses validation', async () => {
    mockIsValid.value = false;

    await wrapper.find(form).trigger('submit');

    expect(spyCreateExercise).toBeCalledTimes(0);

    mockIsValid.value = true;
  });

  it('creates exercise', async () => {
    const CHECKBOX_INDEX = 1;

    expect(spyCreateExercise).toBeCalledTimes(0);
    expect(spyRefetchQueries).toBeCalledTimes(0);
    expect(spyToastSuccess).toBeCalledTimes(0);
    expect(spyRouterPush).toBeCalledTimes(0);

    await wrapper.findComponent(formTitle).setValue(TITLE);
    await wrapper.findComponent(formTitleEn).setValue(TITLE_EN);
    await wrapper.findComponent(formDescription).setValue(DESCRIPTION);
    await wrapper.findComponent(formDescriptionEn).setValue(DESCRIPTION_EN);

    expect(wrapper.find(formIsWeightsRequired).exists()).toBe(false);

    await wrapper.findComponent(formIsWeights).setValue(true);

    expect(wrapper.find(formIsWeightsRequired).exists()).toBe(true);

    await wrapper.findComponent(formIsWeightsRequired).setValue(true);

    await wrapper.findComponent(formEquipmentForWeight).setValue(true);

    wrapper
      .findAllComponents<typeof UiCheckbox>(formMuscleGroups)
      [CHECKBOX_INDEX].vm.$emit('update:modelValue', MUSCLES_FIXTURE[CHECKBOX_INDEX], true);

    await nextTick();

    await wrapper.find(form).trigger('submit');

    expect(spyCreateExercise).toBeCalledTimes(1);
    expect(spyCreateExercise).toBeCalledWith({
      title: TITLE,
      title_en: TITLE_EN,
      description: DESCRIPTION,
      description_en: DESCRIPTION_EN,
      muscles: [MUSCLES_FIXTURE[CHECKBOX_INDEX]],
      equipment: undefined,
      equipmentForWeight: [filterEquipmentByWeights(EQUIPMENTS_FIXTURE, true)[0]],
      isWeights: true,
      isWeightsRequired: true,
    });

    await mockOnSuccess.create?.();

    expect(spyRefetchQueries).toBeCalledTimes(2);
    expect(spyRefetchQueries).toBeCalledWith({ queryKey: [API_EXERCISE] });
    expect(spyRefetchQueries).toBeCalledWith({ queryKey: [API_ACTIVITY_STATISTICS] });

    expect(spyToastSuccess).toBeCalledTimes(1);

    expect(spyRouterPush).toBeCalledTimes(1);
    expect(spyRouterPush).toBeCalledWith(URL_EXERCISE);

    expect(wrapper.emitted('hide')).toHaveLength(1);
  });

  it('updates exercise', async () => {
    expect(spyUpdateExercise).toBeCalledTimes(0);
    expect(spyRefetchQueries).toBeCalledTimes(0);
    expect(spyToastSuccess).toBeCalledTimes(0);

    const NEW_TITLE = 'Новое название';

    await wrapperWithExercise.findComponent(formTitle).setValue(NEW_TITLE);

    await wrapperWithExercise.find(form).trigger('submit');

    expect(spyUpdateExercise).toBeCalledTimes(1);
    expect(spyUpdateExercise).toBeCalledWith({
      _id: EXERCISE_FIXTURE._id,
      title: NEW_TITLE,
      title_en: EXERCISE_FIXTURE.title_en,
      description: EXERCISE_FIXTURE.description,
      muscles: EXERCISE_FIXTURE.muscles,
      createdBy: EXERCISE_FIXTURE.createdBy,
      isWeights: EXERCISE_FIXTURE.isWeights,
      isWeightsRequired: EXERCISE_FIXTURE.isWeightsRequired,
      equipment: EXERCISE_FIXTURE.equipment,
      equipmentForWeight: EXERCISE_FIXTURE.equipmentForWeight,
    });

    await mockOnSuccess.update?.();

    expect(spyRefetchQueries).toBeCalledTimes(2);
    expect(spyRefetchQueries).toBeCalledWith({ queryKey: [API_EXERCISE] });
    expect(spyRefetchQueries).toBeCalledWith({ queryKey: [API_ACTIVITY_STATISTICS] });

    expect(spyToastSuccess).toBeCalledTimes(1);

    expect(wrapper.emitted('hide')).toHaveLength(1);
  });

  it('deletes exercise', async () => {
    expect(spyDeleteExercise).toBeCalledTimes(0);
    expect(spyRemoveQueries).toBeCalledTimes(0);
    expect(spyRefetchQueries).toBeCalledTimes(0);
    expect(spyToastSuccess).toBeCalledTimes(0);
    expect(spyRouterPush).toBeCalledTimes(0);

    wrapperWithExercise.findComponent<typeof FormButtons>(formButtons).vm.$emit('delete', EXERCISE_FIXTURE._id);

    expect(spyDeleteExercise).toBeCalledTimes(1);
    expect(spyDeleteExercise).toBeCalledWith(EXERCISE_FIXTURE._id);

    await mockOnSuccess.delete?.();

    expect(spyRemoveQueries).toBeCalledTimes(1);
    expect(spyRemoveQueries).toBeCalledWith({ queryKey: [API_EXERCISE] });

    expect(spyRefetchQueries).toBeCalledTimes(2);
    expect(spyRefetchQueries).toBeCalledWith({ queryKey: [API_EXERCISE] });
    expect(spyRefetchQueries).toBeCalledWith({ queryKey: [API_ACTIVITY_STATISTICS] });

    expect(spyToastSuccess).toBeCalledTimes(1);

    expect(spyRouterPush).toBeCalledTimes(1);
    expect(spyRouterPush).toBeCalledWith(URL_EXERCISE);

    expect(wrapper.emitted('hide')).toHaveLength(1);
  });

  it('sets form buttons id', async () => {
    expect(wrapper.find(formButtons).attributes('id')).toBe(undefined);
    expect(wrapperWithExercise.find(formButtons).attributes('id')).toBe(EXERCISE_FIXTURE._id);
  });

  it('emits hide by form buttons cancel emit', async () => {
    wrapper.findComponent<typeof FormButtons>(formButtons).vm.$emit('cancel');

    expect(wrapper.emitted('hide')).toHaveLength(1);
  });
});
