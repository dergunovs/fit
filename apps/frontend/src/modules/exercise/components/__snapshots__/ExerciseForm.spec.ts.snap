// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ExerciseForm > matches snapshot 1`] = `
"<div>
  <div tag="form" column="" gap="24" align="flex-start" data-test="exercise-form">
    <fieldset label="Название" isrequired=""><input type="text" modelvalue="" data-test="exercise-form-title"></fieldset>
    <fieldset label="Название EN"><input type="text" modelvalue="" data-test="exercise-form-title-en"></fieldset>
    <fieldset label="Описание">
      <ui-editor-stub modelvalue="" data-test="exercise-form-description"></ui-editor-stub>
    </fieldset>
    <fieldset label="Описание EN">
      <ui-editor-stub modelvalue="" data-test="exercise-form-description-en"></ui-editor-stub>
    </fieldset><input type="checkbox" modelvalue="false" label="Наличие веса у упражнения" data-test="exercise-form-is-weights">
    <!--v-if-->
    <fieldset label="Использованное оборудование"><input type="text" options="[object Object]" isclearable="" islocalefield="false" lang="ru"></fieldset>
    <!--v-if-->
    <div column="">
      <div>Задействованные мышцы</div><input type="checkbox" modelvalue="false" label="Грудь" data-test="exercise-form-muscle-groups"><input type="checkbox" modelvalue="false" label="Ноги" data-test="exercise-form-muscle-groups"><input type="checkbox" modelvalue="false" label="Плечи" data-test="exercise-form-muscle-groups"><input type="checkbox" modelvalue="false" label="Пресс" data-test="exercise-form-muscle-groups"><input type="checkbox" modelvalue="false" label="Руки" data-test="exercise-form-muscle-groups"><input type="checkbox" modelvalue="false" label="Спина" data-test="exercise-form-muscle-groups">
    </div>
    <form-buttons-stub isedit="false" isemitcancel="false" data-test="exercise-form-buttons"></form-buttons-stub>
  </div>
</div>"
`;
